### PR TITLE
EVG-2660 event models for config

### PR DESCRIPTION
**Pull request by john-m-liu**
I'm not finished yet, but wanted to get some eyes on this to see if the model makes sense

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -105,6 +105,19 @@ func (c *AuthConfig) ValidateAndDefault() error {
...
 }
+func (newVal *AuthConfig) GetChanges(before interface{}) (map[string]ConfigDataChange, error) {
+	oldVal, ok := before.(AuthConfig)
+	if !ok {
+		return nil, fmt.Errorf("unable to convert section %s", newVal.SectionId())
+	}
+	changes := map[string]ConfigDataChange{}
+	if newVal.Crowd != nil {
+		if oldVal.Crowd != nil {
+
```
Comment (bsamek) [inline config.go]: missing return

File: config.go [hunk 2]
```diff
@@ -132,6 +145,9 @@ func (c *RepoTrackerConfig) Set() error {
 	return errors.Wrapf(err, "error updating section %s", c.SectionId())
 }
 func (c *RepoTrackerConfig) ValidateAndDefault() error { return nil }
+func (newVal *RepoTrackerConfig) GetChanges(before interface{}) (map[string]ConfigDataChange, error) {
```
Comment (bsamek) [inline config.go]: what is the purpose of these functions that return nil, nil?

File: rest/data/admin.go [hunk 3]
```diff
@@ -52,25 +53,85 @@ func (ac *DBAdminConnector) SetEvergreenSettings(changes *restModel.APIAdminSett
...
+		if !ok {
+			catcher.Add(fmt.Errorf("unable to convert config section %s", propName))
+			continue
+		}
 
-	err = evergreen.SetBanner(text)
+		// save changes in the event log
+		if oldSettings != nil {
+			oldVal := oldStruct.FieldByName(propName)
+			changes, err = section.GetChanges(oldVal.Interface())
```
Comment (bsamek) [inline rest/data/admin.go]: related to our discussion earlier, is there a risk of logging sensitive info the event log?

File: model/event/admin_event.go [hunk 4]
```diff
@@ -1,44 +1,43 @@
...
 )
 
 const (
 	ResourceTypeAdmin = "ADMIN"
-
-	// event types
-	BannerChanged      = "BANNER_CHANGED"
-	ServiceChanged     = "SERVICE_FLAGS_CHANGED"
-	BannerThemeChanged = "THEME_CHANGED"
+	EventType         = "VALUE_CHANGED"
```
Comment (tychoish) [inline model/event/admin_event.go]: maybe `CONFIG_VALUE_CHANGED`

File: config.go [hunk 5]
```diff
@@ -755,6 +870,11 @@ func IsValidBannerTheme(input string) (bool, BannerTheme) {
 	}
 }
 
+type ConfigDataChange struct {
+	Before string `bson:"before" json:"before"`
+	After  string `bson:"after" json:"after"`
```
Comment (tychoish) [inline config.go]: what if before/after aren't strings?

File: model/event/admin_event.go [hunk 6]
```diff
@@ -1,44 +1,43 @@
...
 
 // IsValid checks if a given event is an event on an admin resource
 func (evt AdminEventData) IsValid() bool {
 	return evt.ResourceType == ResourceTypeAdmin
 }
 
-// logAdminEventBase is a helper function to log an admin event
-func logAdminEventBase(eventType string, eventData AdminEventData) error {
-	eventData.ResourceType = ResourceTypeAdmin
+func LogAdminEvent(section string, changes map[string]evergreen.ConfigDataChange, user string) error {
```
Comment (tychoish) [inline model/event/admin_event.go]: does it make sense to move this to a different collection?

**Comment by tychoish** [comment]
LGTM but run the coverage tool and make sure that things are well covered.

**Pull request merged**